colossalai/zero/sharded_optim/sharded_optim_v2.py

```diff
@@ -111,23 +111,46 @@ def __init__(self,
 
         # Store fp32 param shards
         self.master_params: Dict[Parameter, Tensor] = {}
-
-        for group in self.optimizer.param_groups:
-            for p in group['params']:
-                assert hasattr(p, 'col_attr'), 'The parameter must be wrapped with ShardedParam'
-                is_param_sharded = p.col_attr.data.is_sharded
-                if not is_param_sharded:
-                    # TODO (ver217): we may not use shard / gather here
-                    # Param is no sharded, which means we use ZeRO-2 here
-                    # As we only store param shard, we shard it here
-                    self.shard_strategy.shard([p.col_attr.data])
-                self.master_params[p] = cast_tensor_to_fp32(p.col_attr.data.payload).to(self.device)
-                if not is_param_sharded:
-                    # In this branch, there's no need to shard param
-                    # So we gather here
-                    self.shard_strategy.gather([p.col_attr.data])
+        self._first = True
 
     def step(self, *args, **kwargs):
+
+        # When the first time enter the step(), we calculate the margin space for OS
+        # If margin space is enough, part of param fp32 are initialized on GPU.
+        if self._first:
+            self._first = False
+            cuda_margin_space = self.model.cuda_margin_space
+            # div 3 because the space is shared with param fp32, M and V
```

Comment:
What about grad fp32?

Reply:
OK, as we discussed. If we alway keep a grad fp32, it should be div 4 here.

Reply:
ok, I get it.